File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int[][] map;
    static int n;
    static PriorityQueue<Integer> answer;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        map = new int[n][n];
        answer = new PriorityQueue<>();

        for(int i=0; i<n; i++){
            String input = br.readLine();
            for(int j=0; j<n; j++){
                map[i][j] = input.charAt(j) - '0';
            }
        }

        for(int y=0; y<n; y++){
            for(int x=0; x<n; x++){
                if(map[y][x] == 1){
                    bfs(y, x);
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        sb.append(answer.size() + "\n");

        while(!answer.isEmpty()){
            sb.append(answer.poll() + "\n");
        }

        System.out.println(sb);
    }

    public static void bfs(int y, int x){
        int[] dirY = {-1, 0, +1, 0};
        int[] dirX = {0, +1, 0, -1};

        Queue<int[]> que = new LinkedList<>();
        que.add(new int[] {y, x});
        map[y][x] = 0;

        int cnt = 1;

        while(!que.isEmpty()){
            int[] now = que.poll();

            for(int i=0; i<4; i++){
                int newY = now[0] + dirY[i];
                int newX = now[1] + dirX[i];

                if(newY < 0 || newY >= n || newX < 0 || newX >= n) continue;
                if(map[newY][newX] == 0) continue;

                map[newY][newX] = 0;
                que.add(new int[] {newY, newX});
                cnt++;
            }
        }

        answer.add(cnt);
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque

N = int(input())
board = [list(input()) for _ in range(N)]

ans = []

def printBoard():
    for y in range(N):
        for x in range(N):
            print(board[y][x] + ' ', end="")
        print()


def bfs(y, x):
    que = deque()
    que.append([y, x])
    board[y][x] = '0'
    cnt = 0

    dirY = [-1, 0, +1, 0]
    dirX = [0, +1, 0, -1]

    while que:
        nowY, nowX = que.popleft()
        cnt += 1

        for i in range(4):
            newY = nowY + dirY[i]
            newX = nowX + dirX[i]

            if newY < 0 or newY >= N or newX < 0 or newX >= N:
                continue
            if board[newY][newX] == '0':
                continue

            que.append([newY, newX])
            board[newY][newX] = '0'

    return cnt


for y in range(N):
    for x in range(N):
        if board[y][x] == '1':
            cnt = bfs(y, x)
            ans.append(cnt)

print(len(ans))
ans.sort()
for x in ans:
    print(x)